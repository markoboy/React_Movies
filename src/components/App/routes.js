import withSubRoutes from '@components/hocs/WithSubRoutes';
import loadable from '@loadable/component';
import React from 'react';
import { Route } from 'react-router-dom';

export const HOME_PATH = ['/', '/search/:query'];
export const FILM_PATH = ['/film/:id'];

const routes = [
  {
    path: HOME_PATH.concat(...FILM_PATH),
    component: loadable(() => import(/* webpackPreload: true */ '@components/features/PageLayout')),
    exact: true,
    routes: [
      {
        path: HOME_PATH,
        component: loadable(() => import(/* webpackPreload: true */ '@pages/Search')),
        exact: true,
      },
      {
        path: FILM_PATH,
        component: loadable(() => import(/* webpackPrefetch: true */ '@pages/Film')),
        exact: true,
      },
    ],
  },
  {
    path: '*',
    component: loadable(() => import(/* webpackPrefetch: true */ '@pages/NotFound')),
    status: 404,
  },
];

export const getRouteElement = (route) => (
  <Route
    key={`app-route-${route.path}`}
    path={route.path}
    render={withSubRoutes(route)}
    exact={route.exact}
    status={route.status}
  />
);

export default routes;
