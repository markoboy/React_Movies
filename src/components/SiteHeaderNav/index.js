export { default } from './SiteHeaderNav.container';
