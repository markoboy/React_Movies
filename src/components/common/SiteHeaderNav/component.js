import PropTypes from 'prop-types';
import React, { memo } from 'react';
import './styles.scss';

/**
 * @param {Object} props React props
 * @param {import('react').Component} props.brandLogo The header's logo
 * @param {import('react').Component} props.actionButton The header's action button
 * @param {Boolean} props.classes The classes to be applied on the header
 */
function SiteHeaderNavComponent({ headerLogo, actionButton, classes }) {
  return (
    <header className={classes}>
      <div className="header-nav">
        <div className="header-nav__logo-container">{headerLogo}</div>
        <div className="header-nav__btn-container">{actionButton}</div>
      </div>
    </header>
  );
}

SiteHeaderNavComponent.defaultProps = {
  classes: '',
};

SiteHeaderNavComponent.propTypes = {
  headerLogo: PropTypes.element.isRequired,
  actionButton: PropTypes.element.isRequired,
  classes: PropTypes.string,
};

export default memo(SiteHeaderNavComponent);
