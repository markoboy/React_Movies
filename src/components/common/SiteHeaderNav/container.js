import BrandLogo from '@components/common/BrandLogo';
import { MODAL_FORM_ADD_ACTION } from '@components/features/ModalForm/constants';
import clsx from 'clsx';
import PropTypes from 'prop-types';
import React, { useCallback, useMemo } from 'react';
import SiteHeaderNavComponent from './component';
import { HEADER_NAV_CONTAINER_CLASS } from './constants';
import HeaderAddActionButton from './HeaderAddActionButton';
import HeaderSearchActionButton from './HeaderSearchActionButton';

export default function SiteHeaderNavContainer({
  hasBackground,
  onSearch,
  hasSearch,
  setFormAction,
  setIsOpenedModal,
}) {
  const className = useMemo(
    () => (
      clsx(HEADER_NAV_CONTAINER_CLASS, {
        [`${HEADER_NAV_CONTAINER_CLASS}--with-bg`]: hasBackground,
      })
    ),
    [hasBackground]
  );

  const handleHeaderAddClick = useCallback(() => {
    setFormAction(MODAL_FORM_ADD_ACTION);
    setIsOpenedModal(true);
  }, []);

  const actionButton = hasSearch ? (
    <HeaderSearchActionButton onClick={onSearch} />
  ) : (
    <HeaderAddActionButton onClick={handleHeaderAddClick} />
  );

  return (
    <SiteHeaderNavComponent
      headerLogo={<BrandLogo />}
      actionButton={actionButton}
      classes={className}
    />
  );
}

SiteHeaderNavContainer.defaultProps = {
  hasBackground: false,
  hasSearch: false,
};

SiteHeaderNavContainer.propTypes = {
  hasBackground: PropTypes.bool,
  hasSearch: PropTypes.bool,
  onSearch: PropTypes.func.isRequired,
  setFormAction: PropTypes.func.isRequired,
  setIsOpenedModal: PropTypes.func.isRequired,
};
