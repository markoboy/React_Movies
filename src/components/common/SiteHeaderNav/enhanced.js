import { setIsOpenedModalCreator } from '@store/action-creators/modalActionCreators';
import { setFormActionCreator } from '@store/action-creators/modalFormActionCreators';
import { memo } from 'react';
import { connect } from 'react-redux';
import SiteHeaderNavContainer from './container';

const mapDispatchToProps = {
  setFormAction: setFormActionCreator,
  setIsOpenedModal: setIsOpenedModalCreator,
};

export default connect(null, mapDispatchToProps)(memo(SiteHeaderNavContainer));
