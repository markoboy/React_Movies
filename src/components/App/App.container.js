import Spinner from '@components/common/Spinner';
import { LOADING_STATUS } from '@constants/StatusTypes';
import SiteFooter from '@components/SiteFooter';
import SiteHeaderNav from '@components/SiteHeaderNav';
import SiteContainer from '@components/SiteContainer';
import { selectMovieSuccessCreator } from '@store/action-creators/moviesActionCreators';
import { modalIsOpenedSelector } from '@store/selectors/modalSelectors';
import {
  moviesFilterSelector,
  moviesSortBySelector,
  moviesStatusSelector,
  selectedMovieSelector,
} from '@store/selectors/moviesSelectors';
import { fetchMovies } from '@store/thunks/moviesThunk';
import React, {
  lazy,
  Suspense,
  useCallback,
  useEffect,
  useMemo,
} from 'react';
import { connect } from 'react-redux';
import Notification from '@components/Notification';

const LazyModalFormContainer = lazy(() => import('@components/ModalForm'));

function AppContainer({
  sortBy,
  filter,
  selectedMovie,
  status,
  modalIsOpened,
  dispatch,
}) {
  useEffect(() => {
    dispatch(fetchMovies());
  }, [sortBy, filter]);

  const handleOnSearch = useCallback(() => {
    dispatch(selectMovieSuccessCreator(null));
  }, []);

  const showSpinner = useMemo(() => status === LOADING_STATUS, [status]);

  return (
    <>
      <SiteHeaderNav
        hasSearch={!!selectedMovie}
        hasBackground={!!selectedMovie}
        onSearch={handleOnSearch}
      />
      <SiteContainer />

      {showSpinner && <Spinner />}

      {modalIsOpened && (
        <Suspense fallback={<Spinner />}>
          <LazyModalFormContainer />
        </Suspense>
      )}

      <Notification />

      <SiteFooter />
    </>
  );
}

const mapStateToProps = (state) => ({
  sortBy: moviesSortBySelector(state),
  filter: moviesFilterSelector(state),
  selectedMovie: selectedMovieSelector(state),
  status: moviesStatusSelector(state),
  modalIsOpened: modalIsOpenedSelector(state),
});

export default connect(mapStateToProps)(AppContainer);
