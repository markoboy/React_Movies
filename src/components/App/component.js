import SiteFooter from '@components/common/SiteFooter';
import SiteHeaderNav from '@components/common/SiteHeaderNav';
import Spinner from '@components/common/Spinner';
import Notification from '@components/features/Notification';
import PropTypes from 'prop-types';
import React, { lazy, Suspense } from 'react';
import { Switch } from 'react-router-dom';
import { getRouteElement } from './routes';

const LazyModalFormContainer = lazy(() => (
  import(/* webpackPrefetch: true */ '@components/features/ModalForm')
));

function AppComponent({ modalIsOpened, showSpinner, routes }) {
  return (
    <>
      <SiteHeaderNav />

      <Suspense fallback={<Spinner />}>
        <Switch>{routes.map(getRouteElement)}</Switch>

        {modalIsOpened && <LazyModalFormContainer />}
      </Suspense>

      <Notification />

      <SiteFooter />

      {showSpinner && <Spinner />}
    </>
  );
}

AppComponent.propTypes = {
  modalIsOpened: PropTypes.bool.isRequired,
  showSpinner: PropTypes.bool.isRequired,
  routes: PropTypes.arrayOf(PropTypes.object).isRequired,
};

export default AppComponent;
