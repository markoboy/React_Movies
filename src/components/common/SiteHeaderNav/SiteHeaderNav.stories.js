import React from 'react';
import BrandLogo from '../BrandLogo';

import SiteHeaderNav from './component';
import HeaderAddActionButton from './HeaderAddActionButton';
import HeaderSearchActionButton from './HeaderSearchActionButton';

export default {
  component: SiteHeaderNav,
  title: 'Common/SiteHeaderNav',
};

const Template = (args) => <SiteHeaderNav {...args} />;

export const Default = Template.bind({});
Default.args = {
  headerLogo: <BrandLogo />,
};

export const WithBackground = Template.bind({});
WithBackground.args = {
  headerLogo: <BrandLogo />,
  classes: 'header-nav-container--with-bg',
};

export const WithSearchActionButton = Template.bind({});
WithSearchActionButton.args = {
  headerLogo: <BrandLogo />,
  classes: 'header-nav-container--with-bg',
  actionButton: <HeaderSearchActionButton />,
};

export const WithAddActionButton = Template.bind({});
WithAddActionButton.args = {
  headerLogo: <BrandLogo />,
  classes: 'header-nav-container--with-bg',
  actionButton: <HeaderAddActionButton />,
};
