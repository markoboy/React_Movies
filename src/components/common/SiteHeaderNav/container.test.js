import { FILM_PATH, HOME_PATH } from '@components/App/routes';
import { renderWithStore } from '@utils/testUtils';
import React from 'react';
import SiteHeaderNav from './container';

describe('SiteHeaderNav Container', () => {
  it('renders add action button on home page', () => {
    const { container } = renderWithStore(<SiteHeaderNav />, {
      route: HOME_PATH[0],
    });

    expect(container).toMatchSnapshot();
  });

  it('renders search action button on movie detail page', () => {
    const { container } = renderWithStore(<SiteHeaderNav />, {
      route: FILM_PATH[0],
    });

    expect(container).toMatchSnapshot();
  });
});
