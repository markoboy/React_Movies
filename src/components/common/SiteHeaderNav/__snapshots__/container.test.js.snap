// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SiteHeaderNav Container renders add action button on home page 1`] = `
<div>
  <header
    class="header-nav-container"
  >
    <div
      class="header-nav"
    >
      <div
        class="header-nav__logo-container"
      >
        <h1
          class="brand-logo"
        >
          <a
            href="/"
          >
            netflix
            <span
              class="font--regular"
            >
              roulette
            </span>
          </a>
        </h1>
      </div>
      <div
        class="header-nav__btn-container"
      >
        <button
          class="btn btn--secondary btn--with-icon"
          tabindex="0"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-plus fa-w-14 fa-xs "
            data-icon="plus"
            data-prefix="fas"
            focusable="false"
            role="img"
            viewBox="0 0 448 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M416 208H272V64c0-17.67-14.33-32-32-32h-32c-17.67 0-32 14.33-32 32v144H32c-17.67 0-32 14.33-32 32v32c0 17.67 14.33 32 32 32h144v144c0 17.67 14.33 32 32 32h32c17.67 0 32-14.33 32-32V304h144c17.67 0 32-14.33 32-32v-32c0-17.67-14.33-32-32-32z"
              fill="currentColor"
            />
          </svg>
          Add Movie
        </button>
      </div>
    </div>
  </header>
</div>
`;

exports[`SiteHeaderNav Container renders search action button on movie detail page 1`] = `
<div>
  <header
    class="header-nav-container header-nav-container--with-bg"
  >
    <div
      class="header-nav"
    >
      <div
        class="header-nav__logo-container"
      >
        <h1
          class="brand-logo"
        >
          <a
            href="/"
          >
            netflix
            <span
              class="font--regular"
            >
              roulette
            </span>
          </a>
        </h1>
      </div>
      <div
        class="header-nav__btn-container"
      >
        <a
          class="color--primary"
          href="/"
          title="Search movies"
        >
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-search fa-w-16 fa-lg fa-rotate-90 "
            data-icon="search"
            data-prefix="fas"
            focusable="false"
            role="img"
            viewBox="0 0 512 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M505 442.7L405.3 343c-4.5-4.5-10.6-7-17-7H372c27.6-35.3 44-79.7 44-128C416 93.1 322.9 0 208 0S0 93.1 0 208s93.1 208 208 208c48.3 0 92.7-16.4 128-44v16.3c0 6.4 2.5 12.5 7 17l99.7 99.7c9.4 9.4 24.6 9.4 33.9 0l28.3-28.3c9.4-9.4 9.4-24.6.1-34zM208 336c-70.7 0-128-57.2-128-128 0-70.7 57.2-128 128-128 70.7 0 128 57.2 128 128 0 70.7-57.2 128-128 128z"
              fill="currentColor"
            />
          </svg>
        </a>
      </div>
    </div>
  </header>
</div>
`;
